src/Microsoft.Azure.SignalR.AspNet/EndpointProvider/ServiceEndpointProvider.cs

```diff
@@ -76,46 +75,41 @@ public Task<string> GenerateServerAccessTokenAsync(string hubName, string userId
             }
 
             var audience = $"{_endpoint}/{ServerPath}/?hub={GetPrefixedHubName(_appName, hubName)}";
-
             return _accessKey.GenerateAccessTokenAsync(audience, claims, lifetime ?? _accessTokenLifetime, _algorithm);
         }
 
         public string GetClientEndpoint(string hubName = null, string originalPath = null, string queryString = null)
         {
-            var queryBuilder = new StringBuilder();
-
-            if (!string.IsNullOrEmpty(queryString))
+            var uriBuilder = new UriBuilder(_clientEndpoint)
             {
-                queryBuilder.Append(queryString);
-            }
+                Path = ClientPath
+            };
+
+            var queryBuilder = new StringBuilder(queryString?.TrimStart('?'));
 
             if (!string.IsNullOrEmpty(originalPath))
             {
-                if (queryBuilder.Length == 0)
-                {
-                    queryBuilder.Append("?");
-                }
-                else
+                if (queryBuilder.Length != 0)
                 {
                     queryBuilder.Append("&");
                 }
-
                 queryBuilder
                     .Append(Constants.QueryParameter.OriginalPath)
                     .Append("=")
                     .Append(WebUtility.UrlEncode(originalPath));
```

Comment:
shall we use `Uri.EscapeDataString` according to https://stackoverflow.com/a/37110034/687394?
cc @vwxyzh

Reply:
How about reusing `QueryBuilder` https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.extensions.querybuilder?view=aspnetcore-5.0